estate/models/estate_property.py

```diff
@@ -0,0 +1,152 @@
+    facades = fields.Integer("Number of facades", required=True)
+    garage = fields.Boolean("Has a garage", required=True)
+    garden = fields.Boolean("Has a garden", required=True)
+    garden_area = fields.Integer("Garden (sqm)")
+    garden_orientation = fields.Selection(
+        string="Garden orientation",
+        selection=[
+            ("north", "North"),
+            ("west", "West"),
+            ("east", "East"),
+            ("south", "South"),
+        ],
+    )
+    property_type_id = fields.Many2one("estate.property.type", string="Type")
+
+    total_area = fields.Float("Total area (sqm)", compute="_compute_total_area")
+
+    buyer_id = fields.Many2one("res.partner", string="Buyer", copy=False)
+    salesperson_id = fields.Many2one(
+        "res.users",
+        string="Salesperson",
+        default=lambda self: self.env.user,
+    )
+    offer_ids = fields.One2many("estate.property.offer", "property_id", string="Offers")
+    best_offer = fields.Float("Best Offer", compute="_compute_best_offer")
+
+    _sql_constraints = [
+        (
+            "check_expected_price",
+            "CHECK (expected_price > 0)",
+            "The expected price should be positive.",
+        ),
+        (
+            "check_selling_price",
+            "CHECK (selling_price > 0 OR NOT (state = 'sold'))",
+            "The selling price should be positive if property is sold.",
+        ),
+    ]
+
+    @api.depends("living_area", "garden_area")
+    def _compute_total_area(self):
+        for record in self:
+            record.total_area = record.living_area + record.garden_area
+
+    @api.depends("offer_ids")
+    def _compute_best_offer(self):
+        for record in self:
+            record.best_offer = (
+                0
+                if len(record.offer_ids) == 0
+                else max(record.mapped("offer_ids.price"))
+            )
+
+    @api.onchange("garden")
+    def _onchange_garden(self):
+        if self.garden:
+            if self.garden_area == 0:
+                self.garden_area = 10
+            if self.garden_orientation == False:
+                self.garden_orientation = "north"
+        else:
+            self.garden_area = 0
+            self.garden_orientation = None
+
+    @api.onchange("garden_area")
+    def _onchange_garden_area(self):
+        if self.garden_area <= 0:
+            self.garden_area = 0
+            self.garden = False
+            self.garden_orientation = False
+        else:
+            self.garden = True
+            if self.garden_orientation == False:
+                self.garden_orientation = "north"
+
+    def action_set_sold(self):
+        flag = False
+        for record in self:
+            if record.state != "cancelled":
+                record.state = "sold"
+            else:
+                flag = True
+        if flag:
+            raise UserError("Cancelled properties can't be sold.")
+        return True
```

Comment:
```suggestion
    if self.filtered(lambda record: record.state == 'cancelled'):
        raise UserError("Cancelled properties can't be sold.")
        
    self.state = "sold"
```

1. If a `UserError` is raised, the whole transactions will rollback.
2. We don't need to write record by record, we can write in batch by using self directly.

(This should do the same.)

Reply:
This is way better, thanks